src/core/functions/internal_functions/system/MultiSelectModal.ts

```diff
@@ -0,0 +1,229 @@
+import { Modal, TextComponent, fuzzySearch, prepareQuery } from "obsidian";
+import { TemplaterError } from "utils/Error";
+
+export class MultiSelectModal<T> extends Modal {
+    private resolve: (values: T[]) => void;
+    private reject: (reason?: TemplaterError) => void;
+    private selectedItems = new Set<string>();
+    private mappedItems = new Map<string, T>();
+    private submitted = false;
+
+    private SELECTOR_SELECTED = ".is-selected";
+    private CLASS_SELECTED = "is-selected";
+
+    private suggestionsContainer: HTMLDivElement;
+    private chipsContainer: HTMLDivElement;
+    private textInput: TextComponent;
+
+    constructor(
+        private text_items: string[] | ((item: T) => string),
+        private items: T[],
+        private unrestricted: boolean,
+        title: string,
+        private limit?: number
+    ) {
+        super(app);
+        this.initializeModal(items, title);
+    }
+
+    initializeModal(items: T[], title: string) {
+        // If item isn't of type string or items array is empty, then unrestricted is false
+        if (items.length === 0 || typeof items[0] !== "string") {
+            this.unrestricted = false;
+        }
+
+        // Map each item to its text representation
+        items.forEach((item) =>
+            this.mappedItems.set(this.getItemText(item), item)
+        );
+
+        // Create Elements
+        this.titleEl.setText(title);
+        this.chipsContainer = this.contentEl.createDiv();
+        this.textInput = new TextComponent(this.contentEl);
+        this.suggestionsContainer = this.contentEl.createDiv();
+
+        // Add styles
+        this.modalEl.addClass("templater-multiselect-modal");
+        this.contentEl.addClass("templater-multiselect-container");
+        this.chipsContainer.addClass("templater-multiselect-chips");
+        this.suggestionsContainer.addClass("templater-multiselect-suggestions");
+    }
+
+    onOpen(): void {
+        this.textInput.inputEl.addEventListener(
+            "keydown",
+            (evt: KeyboardEvent) => {
+                switch (evt.key) {
+                    case "Tab":
+                        this.addItemFromSuggestionsOrInputField();
+                        break;
+
+                    case "Backspace":
+                        if (!this.textInput.getValue()) this.removeLastItem();
+                        break;
+
+                    case "ArrowDown":
+                        this.navigateSelection("DOWN");
+                        break;
+
+                    case "ArrowUp":
+                        this.navigateSelection("UP");
+                        break;
+
+                    case "Enter":
```

Comment:
I'm not sure why, but hitting the <kbd>Enter</kbd> key while in the input is also adding a newline in the editor, causing a conflict and the following notice to pop up.

<img width="329" alt="image" src="https://github.com/SilentVoid13/Templater/assets/6936914/ee620539-5127-467b-a15d-aa02d56a1430">

I also feels a bit off to have to click on the input then hit the <kbd>Enter</kbd> key if you're clicking or tapping on options instead of using the keyboard and hitting the <kbd>Tab</kbd> key to select options.

Maybe hitting the <kbd>Enter</kbd> key while the modal is open should close the modal, without having to have the input focused?

Might also be worth adding an "Okay" or "Done" button to make it easier for mobile and touch users.

Reply:
> I also feels a bit off to have to click on the input then hit the Enter key if you're clicking or tapping on options instead of using the keyboard and hitting the Tab key to select options.

I will work on it so that we don't have to click on input field to select options and add "Done" button to make more mobile friendly.

I am trying to figure out how to fix the issue (adds a newline in the editor) but didn't get any results if you have any suggestions please fill free to let me know.

Reply:
I'll try to find some time to figure out why it's adding newlines. I don't think the suggester does this, or if it does, I haven't run into that issue yet.